File: Tutorial/Py/TutorialNlpByteMamba3Ag.py
#!/usr/bin/env python3
"""OA Python Tutorial — Byte-Level Mamba-3 SSM, all-position LM (Autograd).

Python port of Tutorial/Ml/TutorialNlpByteMamba3Ag.cpp. The state-space member of the
NLP fair-comparison suite: the untouched Mamba-3
reference mixer + a flat residual, same all-position task / corpus / dims as the rest.

EXPERIMENTAL: the SSM pair is the least-stable surface in the repo (see Release.md).
Trains at lr 0.003 (0.01 diverges late) and is kept experimental-labeled for 0.7.

Model:     OaEmbedding -> OaMamba3Module (+ flat residual) -> OaLinear
Optimizer: AdamW(lr=0.003) ; Backward: implicit autograd via OaGradientTape

Run:
    OA_PYTHON_BUILD_DIR=Build/Release PYTHONPATH=Tutorial/Py \
      Build/PythonVenv/bin/python Tutorial/Py/TutorialNlpByteMamba3Ag.py
"""

from __future__ import annotations

import math
import os
import tempfile

from _oa_import import core, ml, oa, runtime
import _nlp_common as nlp

VOCAB = nlp.BYTE_VOCAB_SIZE  # 256 — byte vocab family
D_STATE, EXPAND, HEAD_DIM = 32, 2, 16


class Mamba3ByteLM:
    """OaEmbedding -> OaMamba3Module ([B,S,D]) -> flat residual -> OaLinear head."""

    def __init__(self):
        self.embed = ml.OaEmbedding(VOCAB, nlp.D_MODEL)
        self.mamba = ml.OaMamba3Module(
            nlp.D_MODEL, d_state=D_STATE, expand=EXPAND, head_dim=HEAD_DIM,
            n_groups=1, rope_fraction=0.5, mimo=False, mimo_rank=4,
            dt_min=0.001, dt_max=0.1, dt_init_floor=1e-4, a_floor=1e-4,
            outproj_norm=True)
        self.head = ml.OaLinear(nlp.D_MODEL, VOCAB)

        for p in self.parameters():
            p.Data.SetRequiresGrad(True)

    def forward(self, x):
        bs, sl = x.Size(0), x.Size(1)
        emb_flat = self.embed.Forward(x)                       # [B*S, D]
        emb_3d = emb_flat.Reshape([bs, sl, nlp.D_MODEL])       # Mamba3 needs [B, S, D]
        y_3d = self.mamba.Forward(emb_3d)                      # [B, S, D]
        mixed = y_3d.Reshape([bs * sl, nlp.D_MODEL]) + emb_flat.Reshape([bs * sl, nlp.D_MODEL])
        return self.head.Forward(mixed)

    def _mods(self):
        return (("embed", self.embed), ("mamba3", self.mamba), ("head", self.head))

    def parameters(self):
        ps = []
        for _, mod in self._mods():
            ps += list(mod.AllParameterPtrs())
        return ps

    def save(self, prefix):
        for name, mod in self._mods():
            mod.Save(f"{prefix}.{name}.oam")

    def load(self, prefix):
        for name, mod in self._mods():
            mod.Load(f"{prefix}.{name}.oam")


def main():
    runtime.OaInitComputeEngine()
    print("\n" + "=" * 66)
    print("  OA Python Tutorial — Byte Mamba-3 SSM · all-position LM (Autograd)")
    print("=" * 66)
    print(f"Vocab: {VOCAB} bytes · Context: {nlp.CONTEXT_LEN} · DModel: {nlp.D_MODEL} · "
          f"DState: {D_STATE} (EXPERIMENTAL)")

    model = Mamba3ByteLM()
    params = model.parameters()
    n_params = sum(p.Data.NumElements() for p in params)
    print(f"Model: Embed({VOCAB}→{nlp.D_MODEL}) → Mamba3(d_state={D_STATE}) + residual "
          f"→ Linear({nlp.D_MODEL}→{VOCAB})")
    print(f"Params: {n_params}    Optimizer: AdamW(lr=0.003)\n")

    sampler = nlp.NlpAllPositionSampler(nlp.CORPUS, nlp.BATCH)
    opt, initial_loss, last_loss, x, y = nlp.train_all_position(
        model.forward, params, sampler, lr=0.003, timer_name="mamba3_step")

    acc = nlp.accuracy_all_positions(model.forward, x, y, VOCAB)
    gen = nlp.generate_greedy(model.forward, "hello", 48, VOCAB)
    print(f"\nBaseline: ln({VOCAB}) = {math.log(VOCAB):.4f}")
    print(f"Initial loss: {initial_loss:.4f} · Final loss: {last_loss:.4f} · Accuracy: {acc:.1f}%")
    print(f"Generation:\n  Prompt:    'hello'\n  Generated: '{gen}'")

    prefix = os.path.join(tempfile.gettempdir(), "oa_py_byte_mamba3_allpos")
    model.save(prefix)
    reloaded = Mamba3ByteLM()
    reloaded.load(prefix)
    reloaded_acc = nlp.accuracy_all_positions(reloaded.forward, x, y, VOCAB)
    print(f"\nSave/Load round-trip: reloaded accuracy {reloaded_acc:.1f}% "
          f"(trained {acc:.1f}%)")

    assert initial_loss > 0.0
    assert last_loss < initial_loss, "Loss must decrease during training"
    assert acc > 30.0, f"All-position accuracy should exceed 30%, got {acc:.1f}%"
    assert abs(reloaded_acc - acc) < 0.5, "Reloaded model must match trained accuracy"
    print("\n✓ All checks passed")


if __name__ == "__main__":
    main()

File: Tutorial/Ml/Nlp/TutorialNlpByteMamba3Ag.cpp
// ═══════════════════════════════════════════════════════════════════════════
// OA Tutorial — Mamba-3 Byte LM (Autograd Reference)
//
// Untouched Mamba-3 reference path: OaEmbedding + OaMamba3Module::Forward()
// (Ssm/Mamba3/Mamba3Siso* kernels) + flat token residual + Linear head.
//
// Does NOT go through OaEmpyrealmCore or Empyrealm-branded shaders.
// Siblings:
//   TutorialNlpByteEmpyrealmAg — Empyrealm fused/traceable SSM (OaEmpyrealmCore)
//   TutorialNlpByteRnnAg / TutorialNlpByteGruAg — recurrent baselines
//   TutorialNlpByteTransformerAg — causal self-attention baseline
// ═══════════════════════════════════════════════════════════════════════════

#include "OaTest.h"
#include "TutorialMl.h"
#include "TutorialNlpCommon.h"
#include <Oa/Ml.h>
#include <Oa/Ml/Byte.h>
#include <Oa/Ml/Autograd.h>
#include <Oa/Runtime/Context.h>
#include <Oa/Runtime/Engine.h>
#include <Oa/Runtime/GpuTimer.h>
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstring>
#include <vector>

static constexpr OaI32 kVocabSize  = OA_BYTE_VOCAB_SIZE;  // 256 — byte vocab family
// kContextLen / kDModel come from TutorialNlpCommon.h (shared suite dims).
static constexpr OaI32 kDState     = 32;
static constexpr OaI32 kExpand     = 2;
static constexpr OaI32 kHeadDim    = 16;

// Direct Mamba-3 reference wiring (no OaEmpyrealmCore).
class OaMamba3ByteLM : public OaModule {
public:
	OaMamba3ByteLM() {
		Embed_ = OaMakeSharedPtr<OaEmbedding>(kVocabSize, kDModel);
		Mamba_ = OaMakeSharedPtr<OaMamba3Module>(
			kDModel, kDState, kExpand, kHeadDim,
			/*NGroups*/ 1, /*RopeFraction*/ 0.5f, /*Mimo*/ false, /*MimoRank*/ 4,
			/*DtMin*/ 0.001f, /*DtMax*/ 0.1f, /*DtInitFloor*/ 1e-4f, /*AFloor*/ 1e-4f,
			/*OutprojNorm*/ true);
		Head_ = OaMakeSharedPtr<OaLinear>(kDModel, kVocabSize);
		RegisterModule("embed", Embed_);
		RegisterModule("mamba3", Mamba_);
		RegisterModule("head", Head_);
	}

	OaMatrix Forward(const OaMatrix& InTokens) override {
		auto bs    = static_cast<OaI64>(InTokens.Size(0));
		auto sl    = static_cast<OaI64>(InTokens.Size(1));
		auto embFlat = Embed_->Forward(InTokens);             // flat [B*S, D]
		auto emb3d   = embFlat.Reshape(OaMatrixShape{bs, sl, kDModel});  // Mamba3 needs [B, S, D]
		auto y3d   = Mamba_->Forward(emb3d);     // [B, S, D] via Mamba3Siso*
		auto mixed = y3d.Reshape(OaMatrixShape{bs * sl, kDModel})
		           + embFlat.Reshape(OaMatrixShape{bs * sl, kDModel});
		return Head_->Forward(mixed);
	}

	void ResetGenerationState(OaI32 InBatch) {
		Mamba_->ResetState(InBatch);
	}

	OaMatrix ForwardGenerationStep(const OaMatrix& InToken) {
		auto batch = static_cast<OaI64>(InToken.Size(0));
		auto embedded = Embed_->Forward(InToken)
			.Reshape(OaMatrixShape{batch, 1, kDModel});
		auto sequenceOutput = Mamba_->Step(embedded);
		return Head_->Forward(
			sequenceOutput.Reshape(OaMatrixShape{batch, kDModel}) +
			embedded.Reshape(OaMatrixShape{batch, kDModel}));
	}

	[[nodiscard]] OaSharedPtr<OaMamba3Module> Mamba() const { return Mamba_; }

private:
	OaSharedPtr<OaEmbedding>     Embed_;
	OaSharedPtr<OaMamba3Module>  Mamba_;
	OaSharedPtr<OaLinear>        Head_;
};

TEST(TutorialNlpByteMamba3Ag, Mamba3Reference) {
	printf("\n╔══════════════════════════════════════════════════════════════════╗\n");
	printf("║  OA Tutorial — Mamba-3 Reference (Autograd)                      ║\n");
	printf("║  OaMamba3Module::Forward · Ssm/Mamba3/ kernels (untouched)       ║\n");
	printf("╚══════════════════════════════════════════════════════════════════╝\n\n");

	OaFnMatrix::SetRngSeed(OaNlpSuiteRngSeed);
	auto  model  = OaMakeSharedPtr<OaMamba3ByteLM>();
	auto  params = model->AllParameterPtrs();
	// lr 0.003: with the corrected [B,S,D] batching the model reaches ~0.3 CE by
	// ~step 120; 0.01 was hot enough to diverge late once batching was fixed.
	auto  opt    = OaMakeUniquePtr<OaAdamW>(params, 0.003F);

	printf("Model: OaEmbedding + OaMamba3Module + flat residual + Linear head\n");
	printf("Params: %lld\n\n", static_cast<long long>(model->NumParameters()));

	constexpr OaI32 kSteps = 300;
	constexpr OaI32 kBatch = 64;
	NlpAllPositionSampler sampler(NlpCorpus(), kBatch);

	TutorialTrainingLoop training(*opt, OaItTrainingConfig{
		.TotalSteps     = kSteps,
		.EpochSteps     = {},
#if defined(_WIN32)
		// The Windows NVIDIA driver loses the device when this long SSM graph
		// has more than one training submission in flight.
#endif
		.BatchSize      = kBatch,
		.SequenceLength = kContextLen,
		.SequenceUnit   = "token",
		.SourceUnitsPerSample = kContextLen,
		.SourceUnit     = "byte",
		.TimerName      = "mamba3_ref_step",
		.Callbacks      = {}
	});

	// Ring-buffered batches: OaItTraining may submit up to MaxAsyncSubmissions()
	// steps before syncing; reusing one batchX/batchY would let the GPU read
	// overwritten token buffers while earlier steps are still in flight.
	auto& ctx = OaContext::GetDefault();
	OaVec<OaMatrix> xRing(static_cast<OaI64>(ctx.MaxAsyncSubmissions()));
	OaVec<OaMatrix> yRing(static_cast<OaI64>(ctx.MaxAsyncSubmissions()));
	OaF32 initialLoss = 0.0F;
	OaF32 lastLoss    = 0.0F;
	float inProjGradL1 = 0.0F;
	float outProjGradL1 = 0.0F;

	while (not training.Loop.IsDone()) {
		const OaI64 step = training.Loop.Index();
		OaMatrix& batchX = xRing[(step - 1) % xRing.Size()];
		OaMatrix& batchY = yRing[(step - 1) % yRing.Size()];
		sampler.NextBatch(batchX, batchY);
		opt->ZeroGrad();
		OaGradientTape tape;
		auto logits = model->Forward(batchX);
		auto loss   = OaFnLoss::CrossEntropy(logits,
			batchY.Reshape(OaMatrixShape{batchY.Size(0) * batchY.Size(1)}));
		tape.Backward(loss);
		training.Loop.Next(loss);

		if (training.Loop.Index() == 1) {
			initialLoss = training.Loop.LiveLoss();
			printf("\n─── Step-1 gradient magnitudes (Mamba3 reference, L1, fp32-read) ───\n");
			auto& ctx2 = OaContext::GetDefault();
			struct MagEntry { const char* name; OaMatrix result; OaI64 numel; };
			std::vector<MagEntry> entries;
			for (auto* p : params) {
				auto g = p->Data.GradMatrix();
				OaMatrix s;
				OaI64 numel = 0;
				if (!g.IsEmpty() && g.NumElements() > 0) {
					auto flat = OaFnMatrix::Cast(g.Reshape(OaMatrixShape{g.NumElements()}), OaScalarType::Float32);
					auto absg = OaFnMatrix::Abs(flat);
					s = OaFnMatrix::Sum(absg, 0);
					numel = p->Data.NumElements();
				}
				entries.push_back({p->Name.CStr(), std::move(s), numel});
			}
			(void)ctx2.Execute();
			(void)ctx2.Sync();
			for (const auto& e : entries) {
				float mag = 0.0F;
				if (e.result.NumElements() > 0) mag = e.result.At(0);
				printf("  %-32s  L1=%.6g  (numel=%lld)\n",
					e.name, mag, static_cast<long long>(e.numel));
				if (std::strcmp(e.name, "in_proj") == 0) inProjGradL1 = mag;
				if (std::strcmp(e.name, "out_proj") == 0) outProjGradL1 = mag;
			}
			printf("\n");
			fflush(stdout);
		}
	}
	ASSERT_TRUE(training.Loop.Finish().IsOk());
	lastLoss = training.Loop.LastLoss();

	OaMatrix& evalX = xRing[(kSteps - 1) % xRing.Size()];
	OaMatrix& evalY = yRing[(kSteps - 1) % yRing.Size()];
	const OaF32 finalBatchAcc = NlpAccuracyAllPositions(*model, evalX, evalY, kVocabSize);
	printf("\nEvaluation:\n");
	printf("  Random-loss baseline ln(%d) = %.4f\n",
		kVocabSize, std::log(static_cast<double>(kVocabSize)));
	printf("  Bits/byte: %.4f\n", NlpBitsPerByte(lastLoss));
	printf("  Accuracy: %.1f%%\n", finalBatchAcc);

	const OaString generated = NlpGenerateStatefulGreedy(
		*model, kNlpGenerationPrompt, kNlpGenerationBytes, kVocabSize);
	printf("\nGeneration:\n");
	printf("  Prompt: '%s'\n", kNlpGenerationPrompt);
	printf("  Generated: '%s'\n\n", generated.c_str());

	EXPECT_LT(lastLoss, initialLoss);
	EXPECT_GT(finalBatchAcc, 30.0F);
	EXPECT_GT(outProjGradL1, 0.0F) << "out_proj must receive gradient on Mamba3 reference path";

	const OaString ckptPath = "/tmp/mamba3_ref_autograd.oam";
	ASSERT_TRUE(model->Save(ckptPath, *opt).IsOk());
	auto reloaded = OaMakeSharedPtr<OaMamba3ByteLM>();
	auto reloadParams = reloaded->AllParameterPtrs();
	auto reloadedOpt = OaMakeUniquePtr<OaAdamW>(reloadParams, 0.003F);
	ASSERT_TRUE(reloaded->Load(ckptPath, *reloadedOpt).IsOk());
	EXPECT_NEAR(NlpAccuracyAllPositions(*reloaded, evalX, evalY, kVocabSize), finalBatchAcc, 1.0F);
	EXPECT_EQ(NlpGenerateStatefulGreedy(
		*reloaded, kNlpGenerationPrompt, kNlpGenerationBytes, kVocabSize), generated);
}
